File: 4.Stack/Kotlin/ValidParentheses.kt
/**
 * Valid Parentheses Problem in Kotlin
 * 
 * This program demonstrates how to check if a string of parentheses is valid.
 * A string is valid if:
 * 1. Open brackets must be closed by the same type of brackets.
 * 2. Open brackets must be closed in the correct order.
 * 3. Every close bracket has a corresponding open bracket of the same type.
 * 
 * Time Complexity: O(n) where n is the length of the string
 * Space Complexity: O(n) in the worst case when all characters are opening brackets
 */

package Kotlin

import java.util.Stack

/**
 * Check if a string of parentheses is valid using a map for matching brackets
 * 
 * @param s The input string containing parentheses
 * @return true if the string is valid, false otherwise
 */
fun isValid(s: String): Boolean {
    // If the string is empty, it's valid
    if (s.isEmpty()) {
        return true
    }
    
    // If the string has odd length, it can't be valid
    if (s.length % 2 != 0) {
        return false
    }
    
    // Map to store matching brackets
    val brackets = mapOf(
        ')' to '(',
        '}' to '{',
        ']' to '['
    )
    
    // Stack to keep track of opening brackets
    val stack = Stack<Char>()
    
    // Iterate through each character in the string
    for (char in s) {
        // If it's a closing bracket
        if (char in brackets.keys) {
            // Check if stack is empty or the top doesn't match
            if (stack.isEmpty() || stack.pop() != brackets[char]) {
                return false
            }
        } else {
            // It's an opening bracket, push to stack
            stack.push(char)
        }
    }
    
    // If stack is empty, all brackets were matched
    return stack.isEmpty()
}

/**
 * Check if a string of parentheses is valid using a switch statement approach
 * 
 * @param s The input string containing parentheses
 * @return true if the string is valid, false otherwise
 */
fun isValidSwitch(s: String): Boolean {
    // If the string is empty, it's valid
    if (s.isEmpty()) {
        return true
    }
    
    // If the string has odd length, it can't be valid
    if (s.length % 2 != 0) {
        return false
    }
    
    // Stack to keep track of opening brackets
    val stack = Stack<Char>()
    
    // Iterate through each character in the string
    for (char in s) {
        when (char) {
            '(', '{', '[' -> stack.push(char)
            ')' -> if (stack.isEmpty() || stack.pop() != '(') return false
            '}' -> if (stack.isEmpty() || stack.pop() != '{') return false
            ']' -> if (stack.isEmpty() || stack.pop() != '[') return false
            else -> return false // Invalid character
        }
    }
    
    // If stack is empty, all brackets were matched
    return stack.isEmpty()
}

/**
 * Print the result of the validation
 * 
 * @param s The input string
 * @param result The validation result
 */
fun printResult(s: String, result: Boolean) {
    println("Input: \"$s\"")
    println("Is valid: $result")
}

/**
 * Generate a long valid string for benchmarking
 * 
 * @param length The desired length of the string
 * @return A valid string of parentheses
 */
fun generateLongValidString(length: Int): String {
    require(length % 2 == 0) { "Length must be even" }
    
    val sb = StringBuilder()
    for (i in 0 until length / 2) {
        sb.append('(')
    }
    for (i in 0 until length / 2) {
        sb.append(')')
    }
    return sb.toString()
}

/**
 * Main function to demonstrate the valid parentheses problem
 */
fun main() {
    println("Valid Parentheses Problem in Kotlin")
    println("----------------------------------")
    
    // Test cases
    val testCases = listOf(
        "()",              // Valid: simple case
        "()[]{}",          // Valid: different types
        "([]){}",          // Valid: nested and adjacent
        "",                // Valid: empty string
        "([",              // Invalid: incomplete
        "([)]",            // Invalid: wrong order
        "(())",            // Valid: nested same type
        "((()()))",        // Valid: complex nesting
        "({}[])",          // Valid: nested different types
        "({[]})",          // Valid: complex nesting different types
        "({",              // Invalid: incomplete
        ")}]",             // Invalid: no opening brackets
        "(((",             // Invalid: only opening brackets
        "((()))",          // Valid: multiple nesting
        "(()())()",        // Valid: multiple groups
        "([)]"             // Invalid: interleaved brackets
    )
    
    // Test using the map approach
    println("\nTesting with Map Approach:")
    for (test in testCases) {
        val result = isValid(test)
        printResult(test, result)
    }
    
    // Test using the switch approach
    println("\nTesting with Switch Approach:")
    for (test in testCases) {
        val result = isValidSwitch(test)
        printResult(test, result)
    }
    
    // Performance comparison
    println("\nPerformance Comparison:")
    val longString = generateLongValidString(100000)
    
    println("Testing with a long valid string of length ${longString.length}")
    
    // Benchmark map approach
    val mapStartTime = System.nanoTime()
    val mapResult = isValid(longString)
    val mapEndTime = System.nanoTime()
    val mapDuration = (mapEndTime - mapStartTime) / 1_000_000.0 // Convert to milliseconds
    
    // Benchmark switch approach
    val switchStartTime = System.nanoTime()
    val switchResult = isValidSwitch(longString)
    val switchEndTime = System.nanoTime()
    val switchDuration = (switchEndTime - switchStartTime) / 1_000_000.0 // Convert to milliseconds
    
    println("Map approach result: $mapResult, Time: $mapDuration ms")
    println("Switch approach result: $switchResult, Time: $switchDuration ms")
    
    if (switchDuration < mapDuration) {
        println("The Switch approach is faster by ${mapDuration - switchDuration} ms")
    } else {
        println("The Map approach is faster by ${switchDuration - mapDuration} ms")
    }
    
    // Functional approach using Kotlin's features
    println("\nFunctional Approach:")
    
    // Extension function for String to check if it has valid parentheses
    fun String.hasValidParentheses(): Boolean = isValid(this)
    
    // Test the extension function
    val validString = "({[]})"  
    val invalidString = "([)]"  
    
    println("\"$validString\" has valid parentheses: ${validString.hasValidParentheses()}")
    println("\"$invalidString\" has valid parentheses: ${invalidString.hasValidParentheses()}")
    
    // Using higher-order functions
    println("\nFiltering valid strings:")
    val validStrings = testCases.filter { it.hasValidParentheses() }
    println("Valid strings: $validStrings")
    
    println("\nGrouping by validity:")
    val groupedByValidity = testCases.groupBy { it.hasValidParentheses() }
    
    println("Valid strings (${groupedByValidity[true]?.size ?: 0}):")
    groupedByValidity[true]?.forEach { println("\"$it\"")}  
    
    println("\nInvalid strings (${groupedByValidity[false]?.size ?: 0}):")
    groupedByValidity[false]?.forEach { println("\"$it\"")}  
}

File: 4.Stack/C++/validParentheses.cpp
/**
 * Valid Parentheses Problem in C++
 * 
 * This program checks if a string containing parentheses, brackets, and braces
 * is valid. A string is valid if:
 * 1. Open brackets must be closed by the same type of brackets.
 * 2. Open brackets must be closed in the correct order.
 * 3. Every close bracket has a corresponding open bracket of the same type.
 * 
 * Time Complexity: O(n) where n is the length of the string
 * Space Complexity: O(n) in the worst case
 */

#include <iostream>
#include <stack>
#include <string>
#include <unordered_map>
#include <vector>

/**
 * Check if a string containing parentheses, brackets, and braces is valid
 * 
 * @param s The input string
 * @return true if the string is valid, false otherwise
 */
bool isValid(const std::string& s) {
    std::stack<char> stack;
    
    // Map to store matching pairs of brackets
    std::unordered_map<char, char> brackets = {
        {')', '('},
        {'}', '{'},
        {']', '['}
    };
    
    for (char c : s) {
        // If current character is an opening bracket, push it to the stack
        if (c == '(' || c == '{' || c == '[') {
            stack.push(c);
        }
        // If current character is a closing bracket
        else if (c == ')' || c == '}' || c == ']') {
            // If stack is empty, there's no matching opening bracket
            if (stack.empty()) {
                return false;
            }
            
            // Check if the top of the stack matches the corresponding opening bracket
            if (stack.top() != brackets[c]) {
                return false;
            }
            
            // Remove the matching opening bracket from the stack
            stack.pop();
        }
    }
    
    // If stack is empty, all brackets are matched
    return stack.empty();
}

/**
 * Alternative implementation using a switch statement instead of a map
 * 
 * @param s The input string
 * @return true if the string is valid, false otherwise
 */
bool isValidSwitch(const std::string& s) {
    std::stack<char> stack;
    
    for (char c : s) {
        switch (c) {
            case '(': case '{': case '[':
                stack.push(c);
                break;
            case ')':
                if (stack.empty() || stack.top() != '(') return false;
                stack.pop();
                break;
            case '}':
                if (stack.empty() || stack.top() != '{') return false;
                stack.pop();
                break;
            case ']':
                if (stack.empty() || stack.top() != '[') return false;
                stack.pop();
                break;
            default:
                // Ignore other characters
                break;
        }
    }
    
    return stack.empty();
}

/**
 * Print the result of the validation
 * 
 * @param s The input string
 * @param result The validation result
 */
void printResult(const std::string& s, bool result) {
    std::cout << "Input: \"" << s << "\"" << std::endl;
    std::cout << "Output: " << (result ? "true" : "false") << std::endl << std::endl;
}

int main() {
    std::cout << "Valid Parentheses Problem Demonstration in C++" << std::endl;
    std::cout << "--------------------------------------------" << std::endl;
    
    // Test cases
    std::vector<std::string> testCases = {
        "()",           // Valid: simple parentheses
        "()[]{}",       // Valid: different types of brackets
        "(]",           // Invalid: mismatched brackets
        "([)]",         // Invalid: incorrect order
        "{[]}",         // Valid: nested brackets
        "",             // Valid: empty string
        "{",            // Invalid: unclosed bracket
        "{[()]}",       // Valid: complex nesting
        "{[()]()}",     // Valid: multiple valid sequences
        "((()))",       // Valid: multiple same brackets
        "((()",         // Invalid: too many opening brackets
        "()))",         // Invalid: too many closing brackets
        "(hello)",      // Valid: with other characters (ignored)
        "[{()}]"        // Valid: complex nesting
    };
    
    // Test using the map-based implementation
    std::cout << "\nUsing Map-based Implementation:" << std::endl;
    std::cout << "-------------------------------" << std::endl;
    for (const auto& test : testCases) {
        bool result = isValid(test);
        printResult(test, result);
    }
    
    // Test using the switch-based implementation
    std::cout << "\nUsing Switch-based Implementation:" << std::endl;
    std::cout << "---------------------------------" << std::endl;
    for (const auto& test : testCases) {
        bool result = isValidSwitch(test);
        printResult(test, result);
    }
    
    // Performance comparison note
    std::cout << "Performance Note:" << std::endl;
    std::cout << "Both implementations have the same time complexity O(n) and space complexity O(n)." << std::endl;
    std::cout << "The switch-based implementation might be slightly faster as it avoids the overhead of hash map lookups." << std::endl;
    std::cout << "The map-based implementation is more maintainable and easier to extend if needed." << std::endl;
    
    return 0;
}